feat: UIG-2639 - vl-cascader - introductie component

Goldflow opened the pull request:
[bamboo](https://www.milieuinfo.be/bamboo/browse/UIGOV-CUWC126)

[jira](https://www.milieuinfo.be/jira/browse/UIG-2639)


[storybook (beta) link](https://milieuinfo.github.io/uigov-builds/develop/1.19.0-develop-cascader-beta.3/storybook/?path=/docs/components-cascader--cascader-default) waar je het in kan bekijken

Goldflow committed:
feat: UIG-2639 - vl-cascader - introductie component

Pieter-Beckers commented on libs/components/src/cascader/stories/vl-cascader.stories-doc.mdx:
```diff
@@ -0,0 +1,132 @@
+
+### Cascader Node - Slots
+
+Bij een `vl-cascader-node`, kan je de slots invullen voor `content` en/of `label`; [zie hier voor meer info](###cascader-node-slots)
```
Eventueel nog een voorbeeld toevoegen hiervoor?

Goldflow commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,311 @@
+     * @param nodeData?: NodeData[]
+     * @private
+     */
+    private updateNavigationStack(label?: string, nodeData?: NodeData[]): void {
```
comment van Pascal
> waarom dan gewoon geen 2 methodes?

Goldflow replied:
ontdubbeld zoals gevraagd

kristof01 commented on apps/playground-lit/src/app/app.element.ts:
```diff
@@ -1,39 +1,145 @@
 export class AppElement extends LitElement {
     static {
-        registerWebComponents([VlPopoverComponent]);
+        registerWebComponents([VlPopoverComponent, VlInfoTile, VlCascaderComponent]);
```
Kan je de registerWebComponents correct zetten?
Dan is het terug een volledig voorbeeld.

kristof01 commented on apps/storybook-e2e/src/e2e/components/cascader/vl-cascader.stories.cy.ts:
```diff
@@ -0,0 +1 @@
+// TODO tests
```
Voeg je hier nog de storybook tests toe?

Goldflow replied:
done

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-arg.ts:
```diff
@@ -0,0 +1,101 @@
+    currentLevel: {
+        name: 'current-level',
+        description:
+            'Huidige niveau in de hiërarchie.\nDit attribuut is reactief en kan je aanpassen om terug te gaan naar een lager niveau.',
```
Ik zou "is reactief" weglaten.

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-arg.ts:
```diff
@@ -0,0 +1,101 @@
+    },
+    isLoading: {
+        name: 'is-loading',
+        description: 'Geeft weer dat data wordt geladen. Is reactief.',
```
Hier zou ik opnieuw "is reactief" weglaten.

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-doc.mdx:
```diff
@@ -0,0 +1,132 @@
+
+# Cascader
+
+De cascader is een component om hiërarchische data weer te geven.
```
Kan je hier de manier hanteren die we gebruiken voor andere componenten?
"Gebruik het 'cascader' component om..."

Goldflow replied:
ok

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-doc.mdx:
```diff
@@ -0,0 +1,132 @@
+
+## Standaard gebruik
+
+Standaard kan je `vl-cascader`-component structuur opbouwen door geneste `vl-cascader-node`-componenten te definiëren. Zie bovenstaande voorbeeld.
```
"kan je de 'cascader' component structuur".
"bovenstaand"

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-doc.mdx:
```diff
@@ -0,0 +1,132 @@
+
+## Inhoud van de nodes aanpassen
+
+Voor bepaalde use-cases zijn kan het belangrijk zijn dat je zelf inhoud kan bepalen.
```
1 keer "zijn" te veel

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-doc.mdx:
````diff
@@ -0,0 +1,132 @@
+<vl-cascader .nodeTemplates=${nodeTemplates}>...</vl-cascader>
+```
+
+Belangrijk als je wil dat in de node verder kan navigeren, dat je in de callback een argument bepaalt van type `NavigateForward`. Deze laat toe om de onderliggende kind-elementen te tonen.
````
"dat je in de node"

kristof01 commented on libs/components/src/cascader/stories/vl-cascader.stories-doc.mdx:
```diff
@@ -0,0 +1,132 @@
+
+### Legacy Documentatie
+
+Dit is een nieuwe component; hiervoor is geen legacy documentatie.
```
"nieuw"

kristof01 commented on libs/components/src/cascader/vl-cascader.component.cy.ts:
```diff
@@ -0,0 +1,103 @@
+
+describe('component vl-cascader', () => {
+    beforeEach(() => {
+        cy.viewport(960, 1440);
```
Is dit nodig?

Goldflow replied:
removed

kristof01 commented on libs/components/src/cascader/vl-cascader.component.cy.ts:
```diff
@@ -0,0 +1,103 @@
+
+registerWebComponents([VlCascaderComponent, VlCascaderNodeComponent]);
+
+const mountDefaultWithArgs = (
```
Je geeft in deze file nooit args mee aan deze methode, je zou de args weg kunnen laten en de mount in de beforeEach doen.

Goldflow replied:
done

kristof01 commented on libs/components/src/cascader/vl-cascader.component.cy.ts:
Kan je ook nog wat tests toevoegen voor de andere manieren waarop je dit component kan gebruiken?

Goldflow replied:
done

kristof01 commented on package.json:
```diff
@@ -148,6 +148,8 @@
         "@govflanders/vl-ui-util": "14.0.2",
+        "@lit-labs/motion": "1.0.4",
+        "@lit-labs/react": "2.1.1",
         "@ungap/custom-elements": "1.2.0",
```
Waarvoor dienen deze 2 packages?

Goldflow replied:
- `@lit-labs/motion`: > verwijderd, die gebruikte ik voorheen om een complexere animatie te voorzien maar dat was uiteindelijk overkill
- `@lit-labs/react`: > gebruikte ik voor react-wrapper in playground-react

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,311 @@
+
+    constructor() {
+        super();
+    }
```
Deze constructor kan weg

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,311 @@
+
+    disconnectedCallback() {
+        super.disconnectedCallback();
+    }
```
Deze disconnectCallback kan weg.

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,311 @@
+    /**
+     * Updates current navigationStack and breadCrumbHistory
+     * If called without both arguments, we go backwards in history
+     * @param label?: string
```
Kan je deze comments vertalen naar het Nederlands?

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,311 @@
+        `;
+    }
+
+    static getTemplateFunctionForType(item: NodeData, cascader: VlCascaderComponent): TemplateResult | void {
```
Je zou deze methode in de utils kunnen zetten ipv een static methode te voorzien.

Goldflow replied:
done

was hier niet zeker over; had het ook overwogen maar dacht gezien hier `TemplateResult` worden teruggegeven, dacht ik dat wel een soort code was die hoorde bij de parent component waarin het wordt gerenderd maar ik kan dit zeker volgen

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,311 @@
+
+    /**
+     * Traverses the DOM tree within VlCascaderComponent,
+     * retrieving the hierarchical data structure of VlCascaderNodeComponent's as an array of NodeData[]
```
Kan je deze comment vertalen naar het Nederlands?

Goldflow replied:
done

kristof01 commented on libs/components/src/cascader/vl-cascader-node.component.ts:
```diff
@@ -0,0 +1,73 @@
+    templateType: string | undefined;
+    requestParams?: unknown;
+    cascaderRef: VlCascaderComponent | undefined;
+    static get properties(): PropertyDeclarations {
```
Blanco lijn toevoegen aub

Goldflow replied:
done

kristof01 commented on libs/components/src/cascader/vl-cascader-node.component.ts:
```diff
@@ -0,0 +1,73 @@
+    constructor() {
+        super();
+        this.label = 'label missing';
+    }
```
Deze default value zet je al vanboven in de file

Goldflow replied:
verwijderd

kristof01 commented on libs/components/src/cascader/vl-cascader-node.component.ts:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    render(): TemplateResult {
+        this.updateItemData();
```
Oppassen met update methodes uit te voeren in de render(), dit kan leiden tot performance hits en reder loops.

Goldflow replied:
verwijderd - moest er niet staan

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,315 @@
+        }
+    };
+
+    static getDefaultNodeTemplate(item: NodeData, cascader: VlCascaderComponent, hasLabelSlot = false): TemplateResult {
```
Deze methode zou je opnieuw bij de utils kunnen steken indien die over meerdere files gebruikt wordt, je ziet maar.

Goldflow replied:
done

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
```diff
@@ -0,0 +1,315 @@
+        `;
+    }
+
+    static defaultNodeActionTemplate(item: NodeData): TemplateResult {
```
Ook eventueel bij de utils.

Goldflow replied:
done

kristof01 commented on libs/components/src/cascader/vl-cascader.component.ts:
Kan je eens nakijken welke methodes er allemaal private zouden kunnen zijn?

Goldflow replied:
done

kspeltix reviewed (commented):
die GetCascaderTemplate en GetCascaderChildren dat zijn callbacks ? Ik zou ze dan niet conform de getter conventie noemen. Ik weet nu ook wel geen super naam: CascaderTemplateCallback / CollectCascaderTemplate / CascaderTemplateCollector / RequestCascaderTemplate / CascaderTemplateRequester / CascaderTemplateFn ?

tonyvandeneynde commented on libs/components/src/cascader/stories/vl-cascader.stories-util.get-children.ts:
```diff
@@ -0,0 +1,45 @@
+import { GetCascaderChildren, NodeData } from '../vl-cascader.model';
+
+export const getCascaderChildren: GetCascaderChildren = async (requestParams: unknown): Promise<NodeData[]> => {
+    // hier kan komt dan een request naar bv. een endpoint
```
Iets te speciale zinsconstructie
hier kan komt dan een request naar bv. een endpoint

Goldflow commented:
> die GetCascaderTemplate en GetCascaderChildren dat zijn callbacks ? Ik zou ze dan niet conform de getter conventie noemen. Ik weet nu ook wel geen super naam: CascaderTemplateCallback / CollectCascaderTemplate / CascaderTemplateCollector / RequestCascaderTemplate / CascaderTemplateRequester / CascaderTemplateFn ?

zoals besproken; naamgeving aangepast

Goldflow committed:
feat: UIG-2539 - customCSS - `custom-css`-attribuut toegevoegd

Voor nieuwe componenten gaan we `data-vl-`-prefix laten vallen. Daarom laten we ook `custom-css` toe als attribuut om een custom CSS string toe te voegen.

Goldflow committed:
fix: UIG-2729 - vl-breadcrumb - breadcrumb styling aangepast conform DV

ksvlaanderen commented on libs/common/storybook/src/stories.helper.ts:
```diff
@@ -76,11 +76,11 @@ export const storyArgs = (args: Args) => {
         ...argTypes,
         customCSS: {
-            name: 'data-vl-custom-css',
+            name: !next ? 'data-vl-custom-css' : 'custom-css',
```
Waar dient dit ! voor?

Goldflow replied:
om de boolean te inverten; had ooit eens gelezen dat bij het verwerken van code de default conditie altijd standaard berekend wordt en je daarom beter voor performance redenen, de meest voorkomende conditie in de `true` conditie moet plaatsen;

ik doe dit sindsdien ook altijd bij de meeste if than else condities; maar even gegoogled en dit blijkt eerder enkel in low level code environments

ksvlaanderen commented on libs/common/utilities/src/base/base.lit.element.ts:
```diff
@@ -3,16 +3,19 @@ import { LitElement, PropertyDeclarations } from 'lit';
     connectedCallback(): void {
         super.connectedCallback();
 
+        this.customCSS = this.customCSSPrefix ? this.customCSSPrefix : this.customCSS;
```
Kan je dit in de addCustomCSS methode doen? Dan vervuil je de connectedCallback niet.

Goldflow replied:
done!

ksvlaanderen commented on package.json:
```diff
@@ -4,14 +4,15 @@
         "test:wctdetail": "npm run build:all && npx nx wctdetail map",
         "test:wctdetailonly": "npx nx wctdetail map",
-        "storybook": "npx nx storybook storybook http://localhost:8080",
+        "storybook": "npx nx storybook storybook http://localhost:8080d",
```
Is dit per ongeluk gebeurd?

Goldflow replied:
dikke vinger - woops (nu gefixt)

ksvlaanderen commented on package.json:
```diff
@@ -4,14 +4,15 @@
     "license": "MIT",
     "scripts": {
         "test": "npm run test:unit && npm run test:wct",
+        "test:component": "npx cypress open --component",
```
Hier zou ik het run commando gebruiken ipv het open, je kan ook een 2de commando voorzien test:component-open bv.

Goldflow replied:
sounds good; is idd beter - aangepast
